Mathlib/Order/SuccPred/Tree.lean

```diff
@@ -0,0 +1,250 @@
+
+/-!
+# Rooted trees
+
+This file proves basic results about rooted trees, represented using the ancestorship order.
+This is a `PartialOrder`, with `PredOrder` with the immediate parent as a predecessor, and an
+`OrderBot` which is the root. We also have an `IsPredArchimedean` assumption to prevent infinite
+dangling chains.
+
+--/
+
+variable {α : Type*} [PartialOrder α] [PredOrder α] [IsPredArchimedean α]
+
+namespace IsPredArchimedean
+
+lemma le_total_of_le {r v₁ v₂ : α} (h₁ : v₁ ≤ r) (h₂ : v₂ ≤ r) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ := by
+  obtain ⟨n, rfl⟩ := h₁.exists_pred_iterate
+  obtain ⟨m, rfl⟩ := h₂.exists_pred_iterate
+  clear h₁ h₂
+  wlog h : n ≤ m
+  · rw [Or.comm]
+    apply this
+    omega
+  right
+  obtain ⟨k, rfl⟩ := Nat.exists_eq_add_of_le h
+  rw [Nat.add_comm, Function.iterate_add, Function.comp_apply]
+  apply Order.pred_iterate_le
+
+lemma lt_or_le_of_le {r v₁ v₂ : α} (h₁ : v₁ ≤ r) (h₂ : v₂ ≤ r) :
+    v₁ < v₂ ∨ v₂ ≤ v₁ := by
+  rw [Classical.or_iff_not_imp_right]
+  intro nh
+  rcases le_total_of_le h₁ h₂ with h | h
+  · apply lt_of_le_of_ne h (ne_of_not_le nh).symm
+  · contradiction
+
+variable [OrderBot α]
+
+section DecidableEq
+
+variable [DecidableEq α]
+
+/--
+The unique atom less than an element in an `OrderBot` with archimedean predecessor.
+-/
+def findAtom (r : α) : α :=
+  Order.pred^[Nat.find (bot_le (a := r)).exists_pred_iterate - 1] r
+
+@[simp]
+lemma findAtom_le (r : α) : findAtom r ≤ r :=
+  Order.pred_iterate_le _ _
+
+@[simp]
+lemma findAtom_bot : findAtom (⊥ : α) = ⊥ := by
+  apply Function.iterate_fixed
+  simp
+
+@[simp]
+lemma pred_findAtom (r : α) : Order.pred (findAtom r) = ⊥ := by
+  unfold findAtom
+  generalize h : Nat.find (bot_le (a := r)).exists_pred_iterate = n
+  cases n
+  · have : Order.pred^[0] r = ⊥ := by
+      rw [← h]
+      apply Nat.find_spec (bot_le (a := r)).exists_pred_iterate
+    simp only [Function.iterate_zero, id_eq] at this
+    simp [this]
+  · simp only [Nat.add_sub_cancel_right, ← Function.iterate_succ_apply', Nat.succ_eq_add_one]
+    rw [← h]
+    apply Nat.find_spec (bot_le (a := r)).exists_pred_iterate
+
+lemma findAtom_ne_bot {r : α} (hr : r ≠ ⊥) :
+    findAtom r ≠ ⊥ := by
+  unfold findAtom
+  intro nh
+  have := Nat.find_min' (bot_le (a := r)).exists_pred_iterate nh
+  replace : Nat.find (bot_le (a := r)).exists_pred_iterate = 0 := by omega
+  simp [this, hr] at nh
+
+@[simp]
+lemma findAtom_eq_bot_iff {r : α} :
+    findAtom r = ⊥ ↔ r = ⊥ where
+  mp h := by_contra fun nh ↦ findAtom_ne_bot nh h
+  mpr h := by simp [h]
+
+@[simp]
+lemma findAtom_is_atom (r : α) (hr : r ≠ ⊥) :
+    IsAtom (findAtom r) := by
+  constructor
+  · simp [hr]
+  · intro b hb
+    apply Order.le_pred_of_lt at hb
+    simpa using hb
+
+end DecidableEq
+
+instance instIsAtomic : IsAtomic α where
+  eq_bot_or_exists_atom_le b := by classical
+    rw [Classical.or_iff_not_imp_left]
+    intro hb
+    use findAtom b, findAtom_is_atom b hb, findAtom_le b
+
+end IsPredArchimedean
```

Comment:
I would move this to an earlier file (and earlier PR!) as well

Reply:
Alright. Together with `coe_pred` that would surely be too long for adding to `SuccPred/Basic` - what would be a good file name for these results?

Reply:
Maybe we should split the archimedean stuff out of `SuccPred.Basic`? then your stuff could go there

Reply:
I like that idea, I always felt the Archimedean stuff to be slightly out of place.

Reply:
I agree regarding `le_total_of_le` and `lt_or_le_of_le`, but `findAtom` is only relevant for trees - I think this is a good place for it

Reply:
Hmm, I would think we want to know that archimedean pred orders are atomic more generally than just for trees, but I guess it can move later